### ISSUE TITLE
Streaming is slow compared to unspecified `chunk_duration`

### ISSUE THREAD

**Issue by eadwu** [OpenMOSS/MOSS-Audio-Tokenizer#11]
The more I reduce the `chunk_duration` the slower it gets. My RTF without specifying a `chunk_duration` is pretty much insanely fast (RTF >> 2) but I specify say the example `0.8` chunk duration and it drops to [eyeballed time but it is bad enough that the minor difference means nothing) ~2x RTF.

**Comment by eadwu** [comment]
The example code is giving me not even real time results

```
            import torch
            from transformers import AutoModel

            repo_id = "OpenMOSS-Team/MOSS-Audio-Tokenizer-v2"
            device = "cuda" if torch.cuda.is_available() else "cpu"
            model = AutoModel.from_pretrained(repo_id, trust_remote_code=True).eval().to(device)
            audio = torch.randn(2, 48000 * 6).to(device)  # dummy stereo waveform

            import time
            start_time = time.perf_counter()
            # 6.0s @ 48kHz = 288000 samples, divisible by downsample_rate=3840
            enc = model.encode(audio.unsqueeze(0), return_dict=True, chunk_duration=0.08)
            # dec = model.decode(enc.audio_codes, return_dict=True, chunk_duration=0.08)
            print(time.perf_counter() - start_time)
```

Roughly on average ~6.8s


In a loop to rule out warmup
```
            import time
            start_time = time.perf_counter()
            # 6.0s @ 48kHz = 288000 samples, divisible by downsample_rate=3840
            enc = model.encode(audio.unsqueeze(0), return_dict=True)
            tock = time.perf_counter()
            enc = model.encode(audio.unsqueeze(0), return_dict=True, chunk_duration=0.08)
            # dec = model.decode(enc.audio_codes, return_dict=True, chunk_duration=0.08)
            print(time.perf_counter() - tock, tock - start_time)
```
giving a 100x difference
```
7.103794400994957 0.6094532120041549
6.942668824995053 0.05837601900566369
6.9523854380022385 0.058620985997549724
6.995213929003512 0.05825569499575067
6.906647559000703 0.06099374100449495
6.860530202000518 0.057195157001842745
6.853343881994078 0.057427323001320474
```

**Comment by CloudRipple** [comment]
Hi, thanks for reporting this!

With `chunk_duration=0.08`, the model encodes the audio sequentially in 80 ms chunks. A 6-second clip requires 75 streaming steps, each maintaining streaming state and incurring overhead from Python scheduling, GPU synchronization, and kernel launches.

Without `chunk_duration`, the model processes the entire clip at once, allowing more parallelism and better amortizing those fixed costs. Lower throughput with smaller chunks is therefore expected.

That said, a gap this large is not necessarily unavoidable. From our experience optimizing MOSS-Audio-Tokenizer streaming decoding in SGLang-Omni, reducing synchronization, reusing buffers, and using CUDA Graphs can help. The gains for the encoding path would need separate validation.

**Comment by eadwu** [comment]
I am trying to set up training data for a streaming TTS model at ~240ms before moving to 80ms chunks. Specifically I am tokenizing a subset of the Emilia dataset. At 1x or even 2x RTF it is not usable for tokenization at all (costs too much).

Without `chunk_duration` it seems to use a default step size of `80ms` but I am not seeing identical results (`enc1 != enc2`) though I don't think the reconstruction sounds that much different.

**Comment by CloudRipple** [comment]
From our experience optimizing MOSS-Audio-Tokenizer, full-sequence and chunked execution do not necessarily produce bitwise-identical results, even when both use FP32. Different input shapes can lead convolutions, matrix multiplications, and attention to use different computation paths or accumulation orders, introducing small floating-point differences.

Once these differences pass through discrete quantization, values close to a decision boundary may produce different codes. So `enc1 != enc2` does not necessarily imply an audible difference in the reconstructed audio.

However, this is only one possible explanation. To determine whether that is the case here, we would need to compare the fraction of differing codes and the differences in the continuous features before quantization, and check that streaming state and chunk boundaries are handled consistently.

**Comment by eadwu** [comment]
Got it. I am seeing that it can be wildly different sometimes (in terms of selecting indices) `(enc1.audio_codes != enc2.audio_codes).float().mean(dim=0)`.

Is unspecified `chunk_duration` still safe for assuming casuality? So no lookahead? I don't particularly see any reason to suspect it would from the paper but I haven't closely checked the code. If so, I'll just use that and 80ms timesteps

```
tensor([[0.0000, 0.0000, 0.0000, 0.1875, 0.0000, 0.0312, 0.0000, 0.0000, 0.2812,
         0.2188, 0.2188, 0.1250, 0.2188, 0.0938, 0.5312, 0.0000, 0.0625, 0.2812,
         0.3125, 0.2188, 0.0312, 0.0938, 0.1250, 0.0000, 0.4062, 0.1875, 0.3750,
         0.3125, 0.5625, 0.1562, 0.3438, 0.2500, 0.7188, 0.5000, 0.2188, 0.5000,
         0.2500, 0.0000, 0.1875, 0.0938, 0.2812, 0.8125, 0.2500, 0.0312, 0.0000,
         0.8125, 0.3438, 0.2500, 0.1562, 0.5000, 0.5312, 0.3750, 0.4688, 0.0312,
         0.8438, 0.8750, 0.8750, 0.3125, 0.0312, 0.1562, 0.9375, 0.2188, 0.2188,
         0.5938, 0.0312, 0.0625, 0.4062, 0.0000, 0.3750, 0.0312, 0.2500, 0.0312,
         0.1250, 0.0938, 0.0000]], device='cuda:0')
```.

**Comment by gyt1145028706** [comment]
Hi @eadwu,

To be clear about what's going on:

**Streaming and non-streaming are mathematically equivalent.** Specifying `chunk_duration` (streaming) and leaving it unspecified (non-streaming / full-sequence) compute exactly the same function. The model is fully causal in both modes — the full-sequence path uses causal convolutions and causal attention masks, so there is **no lookahead** when `chunk_duration` is unspecified. Nothing in the non-streaming path sees future frames. The only difference is *how* the computation is scheduled: one big pass vs. many small passes carrying streaming state.

**The differences you see are numerical error, not a semantic difference.** With different input shapes, cuDNN / cuBLAS pick different convolution and GEMM kernels, and attention uses different tiling — so the accumulation order of floating-point ops changes. This is inherent to GPU computation and cannot be avoided (it's the same reason `torch.matmul` on a full batch vs. row-by-row isn't bitwise identical). These are tiny perturbations in the continuous latent, but discrete quantization is a step function: any latent that lands near a codebook boundary can flip to a neighboring code. Because the model is causal and state is carried forward, the perturbation also propagates through the sequence, which is why your per-frame mismatch rate grows over time. And with residual quantizers, once one layer flips, every subsequent layer sees a different residual and flips too — so "fraction of codes that differ" heavily overstates how different the underlying representations are. Comparing the pre-quantization features or the reconstructed waveforms is a much better measure.

**In practice this has no effect.** From our experience, this level of numerical noise has no measurable impact on the codec's own reconstruction quality, nor on training or inference of downstream models on the tokens. Both modes produce tokens drawn from the same distribution; a downstream model trained on non-streaming tokens will run fine with streaming tokens at inference, and vice versa.

**Recommendation:** for tokenizing a dataset, use unspecified `chunk_duration` — it's causal, orders of magnitude faster, and the tokens are interchangeable with streaming ones. Reserve `chunk_duration` for the actual streaming inference scenario where you don't have the full audio up front.

**Comment by eadwu** [comment]
Got it, thanks for the clarification.
